L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,752 @@
+    vector<TTStubRef> stubs;
+    stubs.reserve(hitPattern.count());
+    for (int layer = 0; layer < setup()->numLayers(); layer++)
+      if (hitPattern[layer])
+        stubs.push_back(stubs_[layer][layerMap[layer]]->ttStubRef());
+    return stubs;
+  }
+
+  // construct TrackKF from Frame
+  TrackKF::TrackKF(const FrameTrack& frame, const DataFormats* dataFormats) :
+    Track(frame, dataFormats, Process::kf) {}
+
+  // construct TrackKF from TrackKfin
+  TrackKF::TrackKF(const TrackKFin& track, double phiT, double qOverPt, double zT, double cot) :
+    Track(track, track.phiT(), track.qOverPt(), track.zT(), track.cot(), track.sectorPhi(), track.sectorEta(), false)
+  {
+    get<0>(data_) += phiT;
+    get<1>(data_) += qOverPt;
+    get<2>(data_) += zT;
+    get<3>(data_) += cot;
+    get<6>(data_) = abs(qOverPt) < dataFormats_->format(Variable::qOverPt, Process::sf).base() / 2. && abs(phiT) < dataFormats_->format(Variable::phiT, Process::sf).base() / 2.;
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // conversion to TTTrack with given stubs
+  TTTrack<Ref_Phase2TrackerDigi_> TrackKF::ttTrack(const vector<StubKF>& stubs) const {
+    TTBV hitVector(0, setup()->numLayers());
+    double chi2phi(0.);
+    double chi2z(0.);
+    vector<TTStubRef> ttStubRefs;
+    const int nLayer = stubs.size();
+    ttStubRefs.reserve(nLayer);
+    for (const StubKF& stub : stubs) {
+      hitVector.set(stub.layer());
+      const TTStubRef& ttStubRef = stub.ttStubRef();
+      chi2phi += pow(stub.phi(), 2) / setup()->v0(ttStubRef, this->qOverPt());
+      chi2z += pow(stub.z(), 2) / setup()->v1(ttStubRef, this->qOverPt());
+      ttStubRefs.push_back(ttStubRef);
+    }
+    static constexpr int nParPhi = 2;
+    static constexpr int nParZ = 2;
+    static constexpr int nPar = nParPhi + nParZ;
+    const int dofPhi = nLayer - nParPhi;
+    const int dofZ = nLayer - nParZ;
+    chi2phi /= dofPhi;
+    chi2z /= dofZ;
+    const int sectorPhi = frame_.first->phiSector();
+    const int sectorEta = frame_.first->etaSector();
+    const double qOverPt = this->qOverPt() / setup()->invPtToDphi();
+    const double phi0 = deltaPhi(this->phiT() + this->qOverPt() * dataFormats_->chosenRofPhi() + dataFormats_->format(Variable::phiT, Process::ht).range() * (sectorPhi - .5));
+    const double cot = this->cot() + setup()->sectorCot(this->sectorEta());
+    const double z0 = this->zT() - this->cot() * setup()->chosenRofZ();
+    static constexpr double trkMVA1 = 0.;
+    static constexpr double trkMVA2 = 0.;
+    static constexpr double trkMVA3 = 0.;
+    const int hitPattern = hitVector.val();
+    const double bField = setup()->bField();
+    TTTrack<Ref_Phase2TrackerDigi_> ttTrack(qOverPt, phi0, cot, z0, chi2phi, chi2z, trkMVA1, trkMVA2, trkMVA3, hitPattern, nPar, bField);
+    ttTrack.setStubRefs(ttStubRefs);
+    ttTrack.setPhiSector(sectorPhi);
+    ttTrack.setEtaSector(sectorEta);
```

Comment:
Setting up eta sector is redundant here because tracklet does not provide any information about eta sector. Suggest either remove this line or feed eta sector information from KF to this TTTrack. ( i.e. this->sectorEta() )

Reply:
I don't understand this comment. There is a field in TTTrack for eta sector and I have an eta sector. The eta sector number is redundant since one can recompute it from cot and z0 but when there is already a data member ...

Reply:
Ah now I understand what your are saying, I am feeding in this->sectorEta().